docs/_ext/codesnippet.py

```diff
@@ -0,0 +1,109 @@
+        "start-after": directives.unchanged_required,
+        "end-before": directives.unchanged_required,
+    }
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.end_marker = None
+        self.location = None
+        self.filename = None
+        self.reader = None
+
+    def run(self) -> List[Node]:
+        self._set_options()
+        self._set_locals()
+        code_snippets = self._get_code_snippets()
+        return [self._make_node(code_snippets)]
+
+    def _get_code_snippets(self) -> List[str]:
+        lines = self.reader.read_file(self.filename, self.location)
+        result = []
+
+        while True:
+            code_snippet, lines = self._get_code_snippet(lines)
+            if not valid_code_snippet(code_snippet):
+                break
+            result.extend(code_snippet)
+        return result
+
+    def _get_code_snippet(self, lines: List[str]) -> Tuple[List[str], List[str]]:
+        """Returns the first code snippet from lines and the rest of lines after that"""
+        try:
+            code_snippet_to_end = self.reader.end_filter(lines, self.location)
+            lines = lines[len(code_snippet_to_end):]
+            code_snippet_start_to_end = self.reader.start_filter(code_snippet_to_end, self.location)
+            code_snippet = self.reader.dedent_filter(code_snippet_start_to_end, self.location)
+        except ValueError as err:
+            if "pattern not found" in str(err):
+                return [], []
+            raise err
+        return code_snippet, lines
+
+    def _set_options(self) -> None:
+        self.options["start-after"] = self.options.get("start-after", self.default_start_marker)
+        self.options["end-before"] = self.options.get("end-after", self.default_end_marker)
+
+    def _set_locals(self) -> None:
+        self.end_marker = self._get_end_marker()
+        self.location = self._get_location()
+        self.filename = self._get_filename()
+        self.reader = LiteralIncludeReader(self.filename, self.options, self.config)
+
+    def _get_end_marker(self) -> str:
+        return self.options["end-before"]
+
+    def _get_location(self) -> Tuple[str, int]:
+        return self.state_machine.get_source_and_line(self.lineno)
+
+    def _get_filename(self) -> str:
+        rel_filename, filename = self.env.relfn2path(self.arguments[0])
+        self.env.note_dependency(rel_filename)
+        return filename
+
+    def _make_node(self, code_snippets) -> Node:
+        text = "".join(code_snippets)
+        node: Element = nodes.literal_block(text, text, source=self.filename)
+
+        if "language" in self.options:
+            node["language"] = self.options["language"]
+        return node
+
+
+def setup(app) -> Dict[str, Any]:
+    app.add_directive("codesnippet", CodeSnippet)
+
+    return {
+        "version": "0.1",
+        "parallel_read_safe": True,
+        "parallel_write_safe": True,
```

Comment:
I'm not sure about this, is it really parallel safe?

Reply:
The docs say:

```
'parallel_write_safe': a boolean that specifies if parallel writing of output files can be used when the extension is loaded. Since extensions usually don’t negatively influence the process, this defaults to True.

Note
  The parallel-write-safe extension must satisfy the following conditions:
  The core logic of the extension is parallelly executable during the writing phase.
```

This extension behaves the same as e.x. literalinclude would and since it is True there, it's default value is True and it works here when set to True I think it should be okay.